File: SpringTILCode/src/test/java/com/example/springtilcode/post/service/PostServiceTest.java
package com.example.springtilcode.post.service;

import com.example.springtilcode.domain.post.controller.dto.response.PostResponse;
import com.example.springtilcode.domain.post.entity.Post;
import com.example.springtilcode.domain.post.exception.PostNotFoundException;
import com.example.springtilcode.domain.post.repository.PostRepository;
import com.example.springtilcode.domain.post.service.PostService;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.security.core.context.SecurityContextHolder;

@SpringBootTest
public class PostServiceTest {
    @Autowired
    private PostService postService;

    @Autowired
    private PostRepository postRepository;

    @BeforeEach
    public void setUser(){
        UserAuthentication userAuthentication = new UserAuthentication("whee050916@gmail.com", null, null);
        SecurityContextHolder.getContext().setAuthentication(userAuthentication);
    }

    @Test
    @DisplayName("1. 글 쓰기")
    public void save(){
        String title = "title 1234";
        String content = "content 1234";

        PostResponse postResponse = postService.save(title, content);

        Assertions.assertEquals(postResponse.getTitle(), title);
        Assertions.assertEquals(postResponse.getContent(), content);
    }

    @Test
    @DisplayName("2. 글 삭제")
    public void delete(){
        Integer id = 1;

        postService.delete(id);

        Assertions.assertTrue(postRepository.findById(id).isEmpty());
    }

    @Test
    @DisplayName("3. 글 보기")
    public void getPost(){
        Integer id = 2;

        PostResponse postResponse = postService.getPost(id);
        Post post = postRepository.findById(id).orElseThrow(PostNotFoundException::new);

        Assertions.assertEquals(post.getTitle(), postResponse.getTitle());
        Assertions.assertEquals(post.getContent(), postResponse.getContent());
    }
}

File: kotlin/src/test/kotlin/com/example/demo/post/service/PostServiceTest.kt
package com.example.demo.post.service

import com.example.demo.domain.post.api.dto.request.PostRequest
import com.example.demo.domain.post.exception.PostNotFoundException
import com.example.demo.domain.post.repository.PostRepository
import com.example.demo.domain.post.service.PostService
import com.example.demo.global.error.GlobalErrorCode
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.context.SecurityContextHolder

@SpringBootTest
class PostServiceTest (
    @Autowired val postService: PostService,
    @Autowired val postRepository: PostRepository
){
    @BeforeEach
    fun setUser(){
        val authentication : Authentication = UsernamePasswordAuthenticationToken("test@gmail.com", "", null)
        SecurityContextHolder.getContext().authentication = authentication
    }

    @Test
    @DisplayName("1. 글 저장")
    fun save(){
        val request = PostRequest("title", "content")

        val postResponse = postService.create(request)

        Assertions.assertEquals(postResponse.title, request.title)
        Assertions.assertEquals(postResponse.content, request.content)
    }

    @Test
    @DisplayName("2. 글 보기")
    fun getPost(){
        val id: Long = 12

        val postResponse = postService.getPost(id)
        val post = postRepository.findById(id).orElseThrow{PostNotFoundException(GlobalErrorCode.BAD_REQUEST)}

        Assertions.assertEquals(postResponse.title, post.title)
        Assertions.assertEquals(postResponse.content, post.content)
        Assertions.assertEquals(postResponse.user?.email, post.user?.email)
    }

    @Test
    @DisplayName("3. 글 삭제")
    fun delete(){
        val id: Long = 3

        postService.delete(id)
        val post = postRepository.findById(id)

        Assertions.assertEquals(post.isEmpty, true)
    }

    @Test
    @DisplayName("4. 글 수정")
    fun editPost(){
        val id: Long = 4
        val request = PostRequest("content1234", "title1234")

        val response = postService.editPost(request, id)

        Assertions.assertEquals(response.title, request.title)
        Assertions.assertEquals(response.content, request.content)
    }
}